Source/FacilityManagementWebsite/SmartHotel.FacilityManagementWeb/SmartHotel.FacilityManagementWeb/ClientApp/src/app/tsi-chart/tsi-chart.component.ts

```diff
@@ -0,0 +1,122 @@
+
+    const dateTimeNowUTC = new Date();
+    const thirtyDaysBack = new Date();
+    thirtyDaysBack.setDate(dateTimeNowUTC.getDate() - Number(environment.tsiHowManyDays));
+
+    const startDate = thirtyDaysBack.toISOString();
+    const endDate = dateTimeNowUTC.toISOString();
+
+    const motionPredicate = this.buildPredicateString(this.motionSensorIds);
+    const lightPredicate = this.buildPredicateString(this.lightSensorIds);
+    const tempPredicate = this.buildPredicateString(this.tempSensorIds);
+
+    const aggregateExpressions = [];
+
+    aggregateExpressions.push(
+      new this.client.ux.AggregateExpression(
+        { predicateString: motionPredicate }, // predicate
+        { property: 'Occupied', type: 'Double' }, // measure column
+        ['avg'], // measure type,
+        { from: startDate, to: endDate, bucketSize: '30m' },  // time range
+        null, // split by value, for you probably just null
+        '#60B9AE', // color
+        'Occupied')  // display name
+    );
+
+    aggregateExpressions.push(
+      new this.client.ux.AggregateExpression(
+        { predicateString: lightPredicate }, // predicate
+        { property: 'Light', type: 'Double' }, // measure column
+        ['avg'], // measure type
+        { from: startDate, to: endDate, bucketSize: '30m' },  // time range
+        null, // split by value, for you probably just null
+        'Green', // color
+        'Light')  // display name
+    );
+
+    aggregateExpressions.push(
+      new this.client.ux.AggregateExpression(
+        { predicateString: tempPredicate }, // predicate
+        { property: 'Temperature', type: 'Double' }, // measure column
+        ['avg'], // measure type
+        { from: startDate, to: endDate, bucketSize: '30m' },  // time range
+        null, // split by value, for you probably just null
+        'Red', // color
+        'Temperature')  // display name
+    );
+
+    const currentThis = this;
+    this.client.server.getAggregates(this.token, environment.tsiFqdn,
+      aggregateExpressions.map(function (ae) { return ae.toTsx(); }))
+      .then(function (result) {
+        const transformedResult = currentThis.client.ux.transformAggregatesForVisualization(result, aggregateExpressions);
+        lineChart.render(transformedResult, { legend: 'compact' }, aggregateExpressions);
+      });
+  }
+
+  ngOnInit() {
+    this.facilityService.executeWhenInitialized(this, this.initializeToken);
+  }
+
+  private initializeToken(self: TsiChartComponent) {
+
+    self.client = new TsiClient();
+    self.token = self.adalService.getCachedToken(environment.tsiApi);
+    if (!self.token) {
+      self.adalService.acquireToken(environment.tsiApi)
+        .subscribe(result => {
+          self.token = result;
+          self.tokenRetrieved = true;
+          self.tryUpdateChart();
+        });
+    } else {
+      self.tokenRetrieved = true;
+      self.tryUpdateChart();
+    }
+  }
+
+  ngOnChanges() {
+    this.tryUpdateChart();
```

Comment:
Should we be utilizing the Angular "differs" to know if there are changes within the collections, or do you think this should catch everything?

Reply:
Good idea - could you point me to example "differes" implementation in the project ?

Reply:
`map.component.ts` has an example "differs" implementation. I think that's the only one